From d34b1e31e5929143dfe8dc6f8928f916ec44af0e Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Fri, 17 Oct 2025 11:13:49 +0200
Subject: [PATCH] bus: stm32_etzpc: allow different access controllers for
 child nodes

Currently, the stm32_etzpc_populate_bus() is only expecting ETZPC
phandles in the access-controllers property, which is improper as
the property is designed to be able to reference multiple access
controllers. Fix it by using the controller referenced by the element
of the access-controllers property to check each access.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I60b98dae135c1121663a3ae35f5fcb0b4e7354c5
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/500634
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 drivers/bus/stm32_etzpc.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/drivers/bus/stm32_etzpc.c b/drivers/bus/stm32_etzpc.c
index 3b6aa421245a..650085d557d5 100644
--- a/drivers/bus/stm32_etzpc.c
+++ b/drivers/bus/stm32_etzpc.c
@@ -107,14 +107,16 @@ static int stm32_etzpc_populate_bus(struct stm32_firewall_controller *ctrl)
 		}
 
 		for (i = 0; i < len; i++) {
-			if (ctrl->grant_access(ctrl, firewalls[i].firewall_id)) {
+			err = firewalls[i].firewall_ctrl->grant_access(firewalls[i].firewall_ctrl,
+								       firewalls[i].firewall_id);
+			if (err) {
 				/*
 				 * Peripheral access not allowed or not defined.
 				 * Mark the node as populated so platform bus won't probe it
 				 */
 				of_detach_node(child);
-				dev_err(parent, "%s: Device driver will not be probed\n",
-					child->full_name);
+				dev_err(parent, "%s: Device driver will not be probed, error: %d\n",
+					child->full_name, err);
 			}
 		}
 
-- 
2.53.0

